formats/opentype.fathom

```diff
@@ -185,6 +186,197 @@ let language_id =
+    data <- match coverage_format {
+        1 => coverage_format_1,
+        2 => coverage_format_2,
+        _ => fail,
+    }
+};
+
+/// # Device and VariationIndex Tables
+///
+/// Device tables and VariationIndex tables are used to provide adjustments to font-unit values in
+/// GPOS, JSTF, GDEF or BASE tables, such as the X and Y coordinates of an attachment anchor
+/// position.
+///
+/// Curiously the table has two interpretations. The second interprtation appears to be have been
+/// tacked on for variable fonts. The gist being that if the delta format is 0x8000 then the table
+/// is a VariationIndex table, which names the fields differently and does not contain a delta
+/// value array. E.g.
+///
+/// let variation_index_table = {
+///     /// A delta-set outer index — used to select an item variation data subtable within the item variation store.
+///     delta_set_outer_index <- u16be,
+///     /// A delta-set inner index — used to select a delta-set row within an item variation data subtable.
+///     delta_set_inner_index <- u16be,
+///     /// Format, = 0x8000
+///     delta_format <- u16be,
+/// };
+///
+/// We only define `device_table` and have it conditionally read the delta value array.
+///
+/// ## References
+///
+/// - [Microsoft's OpenType Spec: Device and VariationIndex Tables](https://docs.microsoft.com/en-us/typography/opentype/spec/chapter2#device-and-variationindex-tables)
+///
+let device_table = (
+    // quotient = numerator / denominator # int division
+    // if quotient * denominator < numerator:
+    //     quotient + 1
+    // else:
+    //     quotient
+    let u16_div_ceil = fun (numerator : U16) => fun (denominator : U16) => (
+        let quotient = u16_div numerator denominator;
+        match (u16_lt (u16_mul quotient denominator) numerator) {
+            true => u16_add quotient 1,
+            false => quotient,
+        }
+    );
+
+    let delta_bits = fun (delta_format : U16) => fun (num_sizes : U16) =>
+        match delta_format {
+            // Signed 2-bit value, 8 values per uint16
+            0x0001 => u16_mul num_sizes 2,
+            // Signed 4-bit value, 4 values per uint16
+            0x0002 => u16_mul num_sizes 4,
+            // Signed 8-bit value, 2 values per uint16
+            0x0003 => u16_mul num_sizes 8,
+            // VariationIndex table, contains a delta-set index pair.
+            0x8000 => 0,
+            _ => 0, // FIXME?
+        };
+
+    let num_sizes = fun (start : U16) => fun (end : U16) =>
+        u16_add (u16_sub end start) 1;
+
+    {
+        /// Smallest size to correct, in ppem
+        start_size <- u16be,
+        /// Largest size to correct, in ppem
+        end_size <- u16be,
+        /// Format of deltaValue array data
+        delta_format <- u16be,
+        /// Array of compressed data
+        _delta_bits <- succeed _ (delta_bits delta_format (num_sizes start_size end_size)),
+        delta_values <- array16 (u16_div_ceil _delta_bits 16) u16be,
```

Comment:
Could maybe use a `let` here:

```
        /// Array of compressed data
        delta_values <- 
            let delta_bits = delta_bits delta_format (num_sizes start_size end_size);
            array16 (u16_div_ceil delta_bits 16) u16be,
```

Reply:
That works!? 🤯 

I've been operating on the assumption that you couldn't use let there and was evidently quick mistaken about #367

Reply:
So the only reason that `succeed` is used here is because that value needs to be used by several fields?

https://github.com/yeslogic/fathom/blob/811673a3833c66942e3282ad1e8c396c05d0ab1b/formats/opentype.fathom#L400

Reply:
Yeah that's right! Introducing shared definitions into a record (which is useful if you don't want to recompute something multiple times, but also just for clarity). I think you can actually kind of think of `succeed` as a bit like the `pure` function in Haskell (aka. `return`). This embeds a ‘pure’ computation in the middle of some binary data.

I've been doing a bit of pondering about this today, thinking some different approaches. A simple way in the short term could be to have something like:

```
let seg_count = u16_div seg_count_x2 2,
```

which would then be elaborated to:

```
seg_count <- succeed _ (u16_div seg_count_x2 2),
```

This still means that it shows up in the resulting record however, which could be a bit of a pain. Still don't have a good solution for this alas.

> I've been operating on the assumption that you couldn't use let there and was evidently quite mistaken about https://github.com/yeslogic/fathom/issues/367

Ohh this might have been a miscommunication. The ‘let format’ in that issue could be seen as ‘effectful’ - in that it lets you do parsing as well. In this case there's no need to do any parsing, so a regular `let` is ok.